docs/DesignDocs/Enabledebugmalloc.md

```diff
@@ -0,0 +1,229 @@
+  until it is terminated and determine if any allocations were leaked.
+  We name it "global detection".
+
+- A developer wants to track allocations in a specific region of the enclave
+  app code that they can specify at dev time. We name it "local detection".
+
+So far the "global detection" has already been fulfilled and improved across
+several pull requests, then this doc will emphasize the design of "local
+detection".
+
+# User Experience
+
+## Global detection
+
+### Default option
+
+The default option is "USE_DEBUG_MALLOC". Now we already set "USE_DEBUG_MALLOC"
+to "ON" in debug mode by default. Users can modify this default option
+according to the need, such as change it to "ON" in release mode or change it
+to "OFF" in debug mode.
+
+### Customed option
+
+Debug malloc will be shipped to the users as a separate library
+oelibdebugmalloc.a, so users can turn on debug malloc per enclave while the
+default option is "OFF".
+
+To use debug malloc, users will have to link oelibdebugmalloc.a into their
+enclaves explictly. This can be done by
+```
+      cmake:
+      target_link_libraries(user-enclave oelibc oedebugmalloc oeenclave)
+
+      make:
+      LDFLAGS=$(shell pkg-config oeenclave-$(COMPILER) --libs)
+      LDFLAGS_CUSTOM=$(subst -loelibc, -loedebugmalloc -loelibc, $(LDFLAGS))
+      $(CC) -o user-enclave user-enc.o $(LDFLAGS_CUSTOM)
+
+```
+
+For the global detection, if there is any memory leak during the enclave's
+life circle, an error code "OE_MEMORY_LEAK" will be raised during the enclave
+termination. This means it not only can check the users' enclave application,
+but also can cover the SDK's allocation of the enclave side.
+
+## Local detection
+
+Currently, debug malloc prints leaks only during enclave termination.
+However, users would like to isolate and track specific memory allocations.
+To achieve this, the following APIs are provided in a new published header `debugmalloc.h`.
+
+- `bool oe_debug_malloc_tracking_start(void)`
+  Start tracking memory allocations.
+
+- `bool oe_debug_malloc_tracking_stop(void)`
+  Stop tracking memory allocations.
+
+- `int oe_debug_malloc_tracking_report(void)`
+  report the number of tracked objects that alive, ie not yet freed.
+
+Please note that the timing and sequences of calling of these functions is
+important:
+
+- The starting and stopping tracking must cover the coupled malloc/free
+operations, which means start tracking before malloc and also stop tracking
+after free, or the report will definitely trigger false alarms.
+
+- oe_debug_malloc_tracking_stop() can only been called after calling
+oe_debug_malloc_tracking_start(), and oe_debug_malloc_tracking_start() can be
+called at the beginning or after calling oe_debug_malloc_tracking_stop(). While
+the calling sequence following this rule, the return values are false. TRUE, or
+non-zero return value means the sequence is wrong and the report does not
```

Comment:
I'd recommend returning oe_result_t instead of bool, so the errors are more meaningful.
E.g. if you call start when it's already started, you might return OE_ALREADY_INITIALIZED or OE_UNEXPECTED.

Reply:
Good suggestion, updated, thanks!